[msal-node]: Add regional authorities telemetry

samuelkubai opened the pull request:
Upgrade the server telemetry schema to version 5.

- Add telemetry on regional authorities

samuelkubai committed:
Change files

samuelkubai committed:
feat: update the confidential and public client test with region discovery metadata

samuelkubai committed:
fix: update the server telemetry manager test

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
fix: update sever telemetry manager test with jest command

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
feat: upgraded the telemetry manager

samuelkubai committed:
fix: client credentials sample

tnorling commented on lib/msal-common/src/authority/Authority.ts:
```diff
@@ -172,6 +176,15 @@ export class Authority {
+     * @returns string
+     */
+    public getRegionDiscoveryMetadata(): RegionDiscoveryMetadata {
+        return this.regionDiscoveryMetadata;
```
Can you make `this.regionDiscoveryMetadata` public instead of adding a getter function?

samuelkubai replied:
Yes, refactored it as suggested.

tnorling commented on lib/msal-common/src/authority/Authority.ts:
```diff
@@ -263,13 +276,32 @@ export class Authority {
 
+                if (this.authorityOptions.azureRegionConfiguration.azureRegion === Constants.AZURE_REGION_AUTO_DISCOVER_FLAG) {
+                    if (autodetectedRegionName) {
+                        this.regionDiscoveryMetadata.region_outcome = RegionDiscoveryOutcomes.AUTO_DETECTION_REQUESTED_SUCCESSFUL;
```
Can we make this a ternary?

samuelkubai replied:
Done

tnorling commented on lib/msal-common/src/authority/Authority.ts:
```diff
@@ -263,13 +276,32 @@ export class Authority {
+                    }
+                } else {
+                    if (autodetectedRegionName) {
+                        if (this.authorityOptions.azureRegionConfiguration.azureRegion === autodetectedRegionName) {
```
Can we also make this a ternary?

samuelkubai replied:
Done

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -34,11 +42,17 @@ export class ServerTelemetryManager {
-        return [SERVER_TELEM_CONSTANTS.SCHEMA_VERSION, request, platformFields].join(SERVER_TELEM_CONSTANTS.CATEGORY_SEPARATOR);
+        const serverTelemetryHeaders = [SERVER_TELEM_CONSTANTS.SCHEMA_VERSION, requestWithRegionDiscoveryFields, platformFields].join(SERVER_TELEM_CONSTANTS.CATEGORY_SEPARATOR);
+        // eslint-disable-next-line no-console
+        console.log(serverTelemetryHeaders);
```
Remove

samuelkubai replied:
Done

samuelkubai committed:
feat: refactor the telemetry implementation

samuelkubai committed:
fix: update the server telemetry documentation

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

tnorling commented on lib/msal-common/src/utils/Constants.ts:
```diff
@@ -349,3 +349,31 @@ export enum  ResponseCodes {
+}
+
+export enum CacheOutcome {
+    NO_CACHE_HIT = "No cache hit",
```
If you make the value of these equal to the number code you can get rid of your switch statement above mapping string to number

samuelkubai replied:
Done, thank you for the suggestion @tnorling

tnorling commented on lib/msal-common/src/utils/Constants.ts:
```diff
@@ -349,3 +349,31 @@ export enum  ResponseCodes {
+ * Region Discovery Outcomes
+ */
+export enum RegionDiscoveryOutcomes {
+    CONFIGURED_MATCHES_DETECTED = "Configured by developer and matches auto-detected",
```
Same here

samuelkubai replied:
Done

samuelkubai committed:
Update change/@azure-msal-common-3150477f-929f-4013-8176-77b1a63f8912.json

Co-authored-by: Sameera Gajjarapu <sameera.gajjarapu@microsoft.com>

samuelkubai committed:
refactor: replace the enum values with the expected header values

samuelkubai committed:
Update change/@azure-msal-node-f63abd0d-4eab-4485-ba86-ca75799c6065.json

Co-authored-by: Sameera Gajjarapu <sameera.gajjarapu@microsoft.com>

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
tests: update the expected telemetry version to 5

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

samuelkubai committed:
test: upgrade the authority definition in the tests

sameerag committed:
Merge branch 'dev' into regional-authorities-telemetry

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -158,4 +165,70 @@ export class ServerTelemetryManager {
+    getRegionSourceValue(regionSource: RegionDiscoverySources | undefined): string {
+        if (!regionSource) return "";
+
+        return regionSource;
```
Why is this function needed? Are we just returning the value passed in?

samuelkubai replied:
Good catch, I have removed it.

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -158,4 +165,70 @@ export class ServerTelemetryManager {
+    getRegionOutcomeValue(regionOutcome: RegionDiscoveryOutcomes | undefined): string {
+        if (!regionOutcome) return "";
+
+        return regionOutcome;
```
Same question here

samuelkubai replied:
Removed this too.

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -158,4 +165,70 @@ export class ServerTelemetryManager {
+     * @returns string
+     */
+    getCacheInfoValue(): string {
+        if (this.forceRefresh) return "1";
```
Isn't forceRefresh already accounted for when setting cacheOutcome? Do we need to duplicate this logic here?

samuelkubai replied:
I have extended the cache outcome enum to include the forceRefresh option and removed this function.

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -158,4 +165,70 @@ export class ServerTelemetryManager {
+    getRegionDiscoveryFields(): string {
+        const regionDiscoveryFields: string[] = [];
+
+        regionDiscoveryFields.push(this.regionUsed ? this.regionUsed : "");
```
```suggestion
        regionDiscoveryFields.push(this.regionUsed || "");
```

sameerag committed:
Update lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts

Co-authored-by: Thomas Norling <thomas.l.norling@gmail.com>

sameerag committed:
Merge branch 'dev' into regional-authorities-telemetry

jasonnutter commented on lib/msal-common/src/client/OnBehalfOfClient.ts:
```diff
@@ -60,6 +60,12 @@ export class OnBehalfOfClient extends BaseClient {
             TimeUtils.isTokenExpired(cachedAccessToken.expiresOn, this.config.systemOptions.tokenRenewalOffsetSeconds)) {
+
+            // Update the server telemetry outcome
+            !cachedAccessToken ?
+                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.NO_CACHED_ACCESS_TOKEN) :
+                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.CACHED_ACCESS_TOKEN_EXPIRED);
```
If you aren't using the returned value, can you use an if statement instead? A little more readable:
```suggestion
            if (!cachedAccessToken) {
                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.NO_CACHED_ACCESS_TOKEN) :
            } else {
                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.CACHED_ACCESS_TOKEN_EXPIRED);
            }
```

samuelkubai replied:
Agreed, both implementations do add to the readability of this code, I have accepted the second version.

jasonnutter commented on lib/msal-common/src/client/OnBehalfOfClient.ts:
```diff
@@ -60,6 +60,12 @@ export class OnBehalfOfClient extends BaseClient {
             TimeUtils.isTokenExpired(cachedAccessToken.expiresOn, this.config.systemOptions.tokenRenewalOffsetSeconds)) {
+
+            // Update the server telemetry outcome
+            !cachedAccessToken ?
+                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.NO_CACHED_ACCESS_TOKEN) :
+                this.serverTelemetryManager?.setCacheOutcome(CacheOutcome.CACHED_ACCESS_TOKEN_EXPIRED);
```
Or this:
```suggestion
            this.serverTelemetryManager?.setCacheOutcome(!cachedAccessToken ? CacheOutcome.CACHED_ACCESS_TOKEN_EXPIRED : CacheOutcome.NO_CACHED_ACCESS_TOKEN);
```

samuelkubai committed:
refactor: improve the readability of server telemetry on the OBO flow client

Co-authored-by: Jason Nutter <janutter@microsoft.com>

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
feat: implemented the changes from the pull request

- bumped up the version for msal-common in the extensions library
- implemented the feedback from the reviewers

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

samuelkubai committed:
Change files

samuelkubai committed:
Update @azure-msal-node-extensions-775467ab-ea19-4849-9a64-291a495d4226.json

samuelkubai committed:
fix: fixing the npm audit issues

samuelkubai committed:
Change files

tnorling commented on lib/msal-common/src/client/SilentFlowClient.ts:
```diff
@@ -60,13 +60,23 @@ export class SilentFlowClient extends BaseClient {
             (cacheRecord.accessToken.refreshOn && TimeUtils.isTokenExpired(cacheRecord.accessToken.refreshOn, 0))) {
+
+            // Update server telemetry manager with the cache outcome
+            if (!cacheRecord.accessToken) {
```
Instead of doing these checks again, can we break down the big if statement above? That would help ensure every condition is covered now and in the future.

samuelkubai replied:
I have broken down this if statement, it's ready for review

tnorling commented on lib/msal-common/src/telemetry/server/ServerTelemetryManager.ts:
```diff
@@ -28,17 +33,21 @@ export class ServerTelemetryManager {
         this.telemetryCacheKey = SERVER_TELEM_CONSTANTS.CACHE_KEY + Separators.CACHE_KEY_SEPARATOR + telemetryRequest.clientId;
+
+        // Initialize the cache outcome value
+        this.setCacheOutcome(this.forceRefresh ? CacheOutcome.FORCE_REFRESH : CacheOutcome.NO_CACHE_HIT);
```
This should be done wherever the cache lookup is done instead (e.g. SilentFlowClient)

tnorling reviewed (commented):
My last suggestions...I promise :) Looks great otherwise!

samuelkubai committed:
fix: implement the changes suggested on the pull request

- improve the code readability
- improve cache outcome initialization

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

samuelkubai committed:
Merge branch 'dev' into regional-authorities-telemetry

samuelkubai committed:
fix: removed unused variable in sever telemetry manager

samuelkubai committed:
Merge branch 'regional-authorities-telemetry' of github.com:AzureAD/microsoft-authentication-library-for-js into regional-authorities-telemetry

samuelkubai committed:
fix: update the server telemetry manager test

sameerag commented on lib/msal-common/src/authority/RegionDiscovery.ts:
```diff
@@ -38,19 +39,27 @@ export class RegionDiscovery {
-                } 
+                }
+                
+                if (autodetectedRegionName) regionDiscoveryMetadata.region_source = RegionDiscoverySources.IMDS;
```
Braces? I do not understand this statement. Is it supposed to be:
```suggestion
                if (autodetectedRegionName) {
                    regionDiscoveryMetadata.region_source = RegionDiscoverySources.IMDS;
                }
```
Same with the below. The whole block isn't very readable.

sameerag commented on lib/msal-common/src/authority/RegionDiscovery.ts:
```diff
@@ -38,19 +39,27 @@ export class RegionDiscovery {
                         return null;
                     }
 
                     const response = await this.getRegionFromIMDS(latestIMDSVersion);
                     if (response.status === ResponseCodes.httpSuccess) {
                         autodetectedRegionName = response.body;
                     }
```
```suggestion
                    const response = await this.getRegionFromIMDS(latestIMDSVersion);
                    if (response.status === ResponseCodes.httpSuccess) {
                        autodetectedRegionName = response.body;
                        regionDiscoveryMetadata.region_source = RegionDiscoverySources.IMDS;
                    }
```

sameerag commented on lib/msal-common/src/authority/RegionDiscovery.ts:
```diff
@@ -38,19 +39,27 @@ export class RegionDiscovery {
-                } 
+                }
+                
+                if (autodetectedRegionName) regionDiscoveryMetadata.region_source = RegionDiscoverySources.IMDS;
```
Get rid of this with the above statement

sameerag replied:
The block is a little fuzzy. We can sync up offline to resolve this.

samuelkubai replied:
I have removed the above statement and refactored the implementation as advised.

sameerag reviewed (changes requested):
Let us clean up the `detectRegion` API a little before we check this in. Everything else looks good.

samuelkubai committed:
fix: cleanup the region discovery api

- improve the readability of the region discovery API

samuelkubai commented:
> Let us clean up the `detectRegion` API a little before we check this in. Everything else looks good.

I have cleaned up the section in question and improved it's readability.

samuelkubai committed:
refactor: improve code readability of the RegionDiscovery file

- improve code readability of the RegionDiscovery file
- add code comments in the RegionDiscovery file

samuelkubai committed:
fix: replace the auto discover flag with the flag from the .NET team

sameerag reviewed (approved):
lgtm
